hadoop-hdds/docs/content/design/degraded-storage-volumes.md

```diff
@@ -0,0 +1,191 @@
+    - Container scanner continues to identify containers that are unhealthy individually
+    - Volume scanner continues to run all checks to determine whether the volume should be failed
+
+##### Failed
+
+Failed volumes are completely inaccessible. We are confident that there is no way to retrieve any container data from this volume, and all replicas on the volume should be reported as missing.
+
+- **Enter**:
+    - **On Startup**:
+        - A RocksDB write handle cannot be acquired. This check is a superset of checking that the volume directory exists.
+    - **While Running**:
+        - Failure threshold of the **failed volume sliding window** is crossed.
+- **Exit**:
+    - **On Startup**:
+        - Entry conditions will be re-evaluated and the volume may become **healthy** or be **failed** again.
+    - **While Running**:
+        - N/A. **failed** is a terminal state while the node is running.
+- **Actions**:
+    - Container scanner stops running
+    - Volume scanner stops running
+    - All containers on the volume are removed from Datanode memory
+    - All containers on the volume are reported missing to SCM
+
+#### Volume Health and Disk Usage
+
+Ideally, volumes that are completely full should not be considered degraded or failed. Just like the rest of the datanode, the scanners will depend on the [VolumeUsage](https://github.com/apache/ozone/blob/2a1a6bf124007821eb68779663bbaad371ea668f/hadoop-hdds/container-service/src/main/java/org/apache/hadoop/ozone/container/common/volume/VolumeUsage.java#L97) API returning reasonably accurate values to assess whether there is enough space to perform checks. **File Check**, for example, must be skipped if there is not enough space to write the file. Additionally, volume checks will be dependent on the space reservation feature to ensure that the RocksDB instance has enough room to write files needed to be opened in write mode and do compactions.
+
+In the future, we may want to handle completely full volumes by opening the RocksDB instance in read-only mode and only allowing read operations on the volume. While this may be a nice feature to have, it is more complicated to implement due to its effects on the delete path. This could have a ripple effect because SCM depends on replicas to ack deletes before it can proceed with operations like deleting empty containers. For simplicity the current proposal considers failure to obtain a DB write handle on startup a full volume failure and delegates read-only volumes as a separate feature.
+
+### Remediating Degraded Volumes
+
+Once degraded volumes can be reported to SCM, it is possible to take action for the affected data on that volume. When a degraded volume is encountered, we should:
+- Stop writing new data onto the volume
+- Make additional copies of the data on that volume, ideally using other nodes as sources if possible
+- Notify the admin when all data from the volume has been fully replicated so it is safe to remove
```

Comment:
Another possibly simpler option, but not without its own challenges, would be to use a variation of the disk balancer to copy "as much as possible" off the old disk onto other disks on the same node. Then fail the disk and let the hopefully few containers with issues be replicated via the normal path.

This does get into just how bad the disk is - if most reads are failing or are very slow, this is unworkable probably, but then also the disk is probably affecting normal cluster read ops.

Reply:
Yes, this shows another way of action to handled degraded volume quickly.

Reply:
Yes this is getting into the decom implementation which is not really in scope for this document. This section is just to demonstrate how degraded volumes could be handled by a generic volume decom feature with only a loose coupling between the two features' implementations.